### PR TITLE
base58: use map instead of strchr() when decode

### PR DESCRIPTION
**Pull request by bitkevin**
Use array map instead of find string position.

Test code snippet:

```cpp

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include <string>

int main(int argc, const char * argv[]) {

  static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  static const int8_t mapBase58[] = {
    -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
    -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
    -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
    -1, 0, 1, 2, 3, 4, 5, 6,  7, 8,-1,-1,-1,-1,-1,-1,
    -1, 9,10,11,12,13,14,15, 16,-1,17,18,19,20,21,-1,
    22,23,24,25,26,27,28,29, 30,31,32,-1,-1,-1,-1,-1,
    -1,33,34,35,36,37,38,39, 40,41,42,43,-1,44,45,46,
    47,48,49,50,51,52,53,54, 55,56,57,-1,-1,-1,-1,-1,
  };

  const std::string b58Str(pszBase58);

  for (size_t i = 0; i < b58Str.length(); i++) {
    const char *ch = strchr(pszBase58, b58Str[i]);
    printf("%d - %d\n", ch - pszBase58, mapBase58[(uint8_t)b58Str[i]]);
    assert(ch - pszBase58 == mapBase58[(uint8_t)b58Str[i]]);
  }

  assert(mapBase58['1'] == 0);
  assert(mapBase58['z'] == 57);

  /** All alphanumeric characters except for "0", "I", "O", and "l" */
  assert(mapBase58['0'] == -1);
  assert(mapBase58['I'] == -1);
  assert(mapBase58['O'] == -1);
  assert(mapBase58['l'] == -1);

  return 0;
}

```

### CODE REVIEW

File: src/base58.cpp [hunk 1]
```diff
@@ -31,11 +41,9 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
     // Process the characters.
     while (*psz && !isspace(*psz)) {
         // Decode base58 character
-        const char* ch = strchr(pszBase58, *psz);
-        if (ch == nullptr)
+        int carry = mapBase58[(uint8_t)*psz];
```
Comment (ken2812221) [inline src/base58.cpp]: Should you check the character `*psz` is less than 128?

**Comment by promag** [comment]
Have you measured performance improvement?

File: src/base58.cpp (hunk 1 above)
Comment (bitkevin) [reply src/base58.cpp]: Oops...I was thought size of `mapBase58` is 256, thanks

**Comment by bitkevin** [comment]
> Have you measured performance improvement?

Performance improvement is about 40%. 1,000,000 rounds, it's about `450ms` vs `250ms`.

```
uint64_t getCurrentTime() {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return tv.tv_sec * 1000 + tv.tv_usec / 1000;
}

const std::string b58Str("3J98t1WpEZ73CNmQviecrnyiWrnqRhWNLy");
size_t len = b58Str.length();

uint64_t cnt;

cnt = 0;
uint64_t t1 = getCurrentTime();
for (size_t j = 0; j < 1000000; j++) {
  for (size_t i = 0; i < len; i++) {
    const char *ch = strchr(pszBase58, b58Str[i]);
    cnt += ch - pszBase58;
  }
}
uint64_t t2 = getCurrentTime();
printf("%lld\n", t2 - t1);

cnt = 0;
uint64_t t3 = getCurrentTime();
for (size_t j = 0; j < 1000000; j++) {
  for (size_t i = 0; i < len; i++) {
    cnt += mapBase58[(uint8_t)b58Str[i]];
  }
}
uint64_t t4 = getCurrentTime();
printf("%lld\n", t4 - t3);
```

**Comment by laanwj** [comment]
Concept ACK. Seems very straightforward (haven't checked the table yet, though).

> Performance improvement is about 40%. 1,000,000 rounds, it's about 450ms vs 250ms.

Nice. Though here you're not benchmarking the entire `DecodeBase58` function, but the specific part that you sped up, so that will give somewhat distored results.

FWIW there's a benchmark for base58 in `src/bench` - it's somewhat representative of what base58 is used for in bitcoin - encoding/decoding addresses.

File: src/base58.cpp [hunk 2]
```diff
@@ -31,11 +41,11 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
     // Process the characters.
     while (*psz && !isspace(*psz)) {
         // Decode base58 character
-        const char* ch = strchr(pszBase58, *psz);
-        if (ch == nullptr)
+        if ((uint8_t)*psz >= sizeof(mapBase58)/sizeof(mapBase58[0]))
```
Comment (laanwj) [inline src/base58.cpp]: Just make the mapBase58 array 256 bytes large, and you don't need the range check.

File: src/base58.cpp [hunk 3]
```diff
@@ -29,13 +47,12 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
     int size = strlen(psz) * 733 /1000 + 1; // log(58) / log(256), rounded up.
     std::vector<unsigned char> b256(size);
     // Process the characters.
+    assert(sizeof(mapBase58)/sizeof(mapBase58[0]) == 256);  // guarantee not out of range
```
Comment (promag) [inline src/base58.cpp]: IMO just leave this as a comment or make it `static_assert`?

File: src/base58.cpp (hunk 1 above)
Comment (promag) [reply src/base58.cpp]: Fixed, as per @laanwj suggestion `mapBase58` has 256 elements.

**Comment by donaloconnor** [comment]
utACK. Your test in comment 0 threw me off because it only has 128 elements.

I think we need tests also submitted in this to test all values 0-255.

**Comment by MarcoFalke** [comment]
Would you mind removing the test in comment 0 and adding it to the unit test suite?

File: src/base58.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/base58.cpp]: With c++11 it should be easy to make it a static_assert.

```diff
--- a/src/base58.cpp
+++ b/src/base58.cpp
@@ -20,7 +20,7 @@
 
 /** All alphanumeric characters except for "0", "I", "O", and "l" */
 static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
-static const int8_t mapBase58[] = {
+constexpr std::array<int8_t, 256> mapBase58{
     -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
     -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
     -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
@@ -55,7 +55,7 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
     int size = strlen(psz) * 733 /1000 + 1; // log(58) / log(256), rounded up.
     std::vector<unsigned char> b256(size);
     // Process the characters.
-    assert(sizeof(mapBase58)/sizeof(mapBase58[0]) == 256);  // guarantee not out of range
+    static_assert(mapBase58.size() == 256); // guarantee not out of range
     while (*psz && !isspace(*psz)) {
         // Decode base58 character
         int carry = mapBase58[(uint8_t)*psz];

**Comment by sipa** [comment]
I benchmarked this on my desktop system: a full address decode goes from 1.14 us to 0.93 us. 

I'm not convinced this is worth it.

**Comment by randolf** [comment]
@sipa One use case I can think of immediately is that Vanity Address Generators can benefit from this performance increase because they repeatedly use the Base58-encoded addresses in attempting to match the desired string(s).

Generally I also value speed optimizations, even in normal application use, and I don't regard the increased size of the resulting binary to be significant enough to warrant trade-off concern here.  YMMV.

**Review by dcousens — APPROVED** [review]
IMHO,  simpler,  utACK.

nit: why is `mapBase58` indented in the 8th element?

File: src/base58.cpp (hunk 3 above)
Comment (laanwj) [reply src/base58.cpp]: A run-time assertion is certainly overkill here.

File: src/base58.cpp [hunk 4]
```diff
@@ -12,6 +12,24 @@
 
 /** All alphanumeric characters except for "0", "I", "O", and "l" */
 static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+static const int8_t mapBase58[] = {
```
Comment (laanwj) [inline src/base58.cpp]: If you make this
```
static const int8_t mapBase58[256] = {
```
That's pretty much a static assertion that the size will be 256.

File: src/base58.cpp [hunk 5]
```diff
@@ -47,11 +47,11 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
     int size = strlen(psz) * 733 /1000 + 1; // log(58) / log(256), rounded up.
     std::vector<unsigned char> b256(size);
     // Process the characters.
-    assert(sizeof(mapBase58)/sizeof(mapBase58[0]) == 256);  // guarantee not out of range
+    static_assert(mapBase58.size() == 256, "mapBase58.size() should be 256"); // guarantee not out of range
```
Comment (dcousens) [inline src/base58.cpp]: The `static_assert` is pointless now?

File: src/base58.cpp [hunk 6]
```diff
@@ -12,7 +12,7 @@
 
 /** All alphanumeric characters except for "0", "I", "O", and "l" */
 static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
-static const int8_t mapBase58[] = {
+static constexpr std::array<int8_t, 256> mapBase58 {
```
Comment (donaloconnor) [inline src/base58.cpp]: missing #include <array> ?

File: src/base58.cpp (hunk 5 above)
Comment (bitkevin) [reply src/base58.cpp]: Just in case, after all `static_assert()` is no harm.

File: src/base58.cpp (hunk 6 above)
Comment (bitkevin) [reply src/base58.cpp]: sorry, I don't familiar with c++11, so just change it  back to old school style.

Comment (dcousens) [reply src/base58.cpp]: IMHO,  the `.size()` method was worth the import... but anyway.

**Comment by promag** [comment]
utACK 5d71e4d, please squash.

**Comment by sipa** [comment]
utACK, needs squash.

**Comment by eklitzke** [comment]
This is good for a 20% speedup for me with GCC 7.3 (median goes from 8.70969e-07 to 7.00866e-07). ACK once squahsed.

**Comment by JeremyRubin** [comment]
I'd like to see a comparison with one or two other methods of doing a table lookup to make sure this is optimal.

For instance

```c++
switch(ch) {
  case '1': 
    carry = 0;
    break;
  // ....
  case 'z':
    carry = 57;
    break;
  default:
    return false;
}
```
Additionally you can try some outputs from gperf https://www.gnu.org/software/gperf/

**Comment by eklitzke** [comment]
I don't think you can get any faster than this approach, which is a flat lookup table that maps ints to ints (without any hashing).

The typical use case of gperf is for something kind of different: you'd provide it to a tokenizer where you have a grammar of long human readable strings, and you want to hash all of the tokens in the grammar to small ints without collisions.

**Comment by sipa** [comment]
Stop wasting time on discussing the performance. This does not matter. Decoding an address could take 50 us and I don't think anyone would notice.

**Comment by JeremyRubin** [comment]
I have some notes on why some alternatives that would be faster, but as @sipa notes, there are bigger fish to fry.

**Comment by laanwj** [comment]
> If the resulting code looks better, go for it. Otherwise, don't.

Yes, I do prefer the code like this, because it's more consistent with how we handle base32 and hex for ex. So utACK. Agree that this is a dead alley in regard to performance, if you are interested in performance please review @eklitzke's work he's doing great things.

**Pull request merged**